### PR TITLE
Self-modifying config.

### PR DESCRIPTION
**Pull request by dkorolev**

### CODE REVIEW

File: Blocks/SelfModifyingConfig/config.h [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    } catch (const current::serialization::json::TypeSystemParseJSONException&) {
+      CURRENT_THROW(SelfModifyingConfigParseJSONException(contents));
+    }
+  }
+
+  // Not thread-safe.
+  void Update(const config_t& new_config, std::chrono::microseconds now = current::time::Now()) {
+    // First, save the curent config under a new name.
+    // This could be atomic rename, but meh. -- D.K.
+    const std::string new_filename = filename_ + current::FormatDateTime(now, ".%Y%m%d-%H%M%S");
```
Comment (sompylasar) [inline Blocks/SelfModifyingConfig/config.h]: Use helper HistoricalFilename?

Comment (dkorolev) [inline Blocks/SelfModifyingConfig/config.h]: :+1: 46a459f

File: Blocks/SelfModifyingConfig/test.cc [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+  current::time::SetNow(std::chrono::microseconds(429865200ll * 1000ll * 1000ll));  // I was born. -- D.K.
+
+  current::SelfModifyingConfig<UnitTestSelfModifyingConfig> config(config_filename);
+  EXPECT_EQ(1, config.Config().x);
+
+  EXPECT_EQ(JSON(UnitTestSelfModifyingConfig(1)), current::FileSystem::ReadFileAsString(config_filename));
+  EXPECT_EQ(JSON(UnitTestSelfModifyingConfig(1)), current::FileSystem::ReadFileAsString(historical_filename));
+
+  current::time::SetNow(std::chrono::microseconds(1439708400ll * 1000ll * 1000ll));  // I turned 32. -- D.K.
```
Comment (sompylasar) [inline Blocks/SelfModifyingConfig/test.cc]: 👍 Could you please explain the 💯 011 magic?

Comment (dkorolev) [inline Blocks/SelfModifyingConfig/test.cc]: it's `0ll`, for `long long`, to force the constant to be 64-bit. Unnecessary in (C++11)++, just a habit.

File: Blocks/SelfModifyingConfig/test.cc [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+  const std::string config_filename =
+      current::FileSystem::JoinPath(FLAGS_self_modifying_config_test_tmpdir, CurrentTestName() + "Config");
+  const auto original_file_remover = current::FileSystem::ScopedRmFile(config_filename);
+
+  const std::string historical_filename = config_filename + ".19830816-000000";
+  const auto historical_file_remover = current::FileSystem::ScopedRmFile(historical_filename);
+
+  current::time::SetNow(std::chrono::microseconds(429865200ll * 1000ll * 1000ll));  // I was born. -- D.K.
+
+  current::FileSystem::WriteStringToFile("De Louboutin.", config_filename.c_str());
```
Comment (sompylasar) [inline Blocks/SelfModifyingConfig/test.cc]: And wearing awesome pants?

Comment (dkorolev) [inline Blocks/SelfModifyingConfig/test.cc]: В Челябинске лучше торчать.

File: Blocks/SelfModifyingConfig/test.cc (hunk 2 above)
Comment (sompylasar) [inline Blocks/SelfModifyingConfig/test.cc]: Oh that... Thanks! Hard to tell "one" from "lowercase el" in this font...

File: Blocks/SelfModifyingConfig/test.cc (hunk 3 above)
Comment (sompylasar) [inline Blocks/SelfModifyingConfig/test.cc]: LMFAO

Comment (dkorolev) [inline Blocks/SelfModifyingConfig/test.cc]: Там было `"Пламенный как\nпионерский привет!"` вначале, но golden output был с юникодом, что не айс.

File: scripts/MakefileImpl [hunk 4]
```diff
@@ -63,6 +63,7 @@ clean:
 	find . -type d -name .current | xargs rm -rf
 	find . -type f -name 'current_build.h' | xargs rm -f
 	find . -type f -name 'current_build.h.*' | xargs rm -f
+	find . -type f -name '.current_std???.txt' | xargs rm -f
```
Comment (sompylasar) [inline scripts/MakefileImpl]: What the `???`?

Comment (dkorolev) [inline scripts/MakefileImpl]: `stdout`, `stderr`. Temporary outputs of a `make check` failed in the middle, which we forgot to remove in `make check`.

File: Blocks/SelfModifyingConfig/config.h [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+SOFTWARE.
+*******************************************************************************/
+
+#ifndef BLOCKS_SELFMODIFYINGCONFIG_CONFIG_H
+#define BLOCKS_SELFMODIFYINGCONFIG_CONFIG_H
+
+// `SelfModifyingConfig<T>` reads the config of type `T`, and can update it dynamically, preserving the history.
+//
+// `SelfModifyingConfig<T>`, takes one required constructor parameter: path to the file from where the config
+// should be read. The underlying config can be accessed via `.Config()`, or a cast to `conts T&`.
```
Comment (mzhurovich) [inline Blocks/SelfModifyingConfig/config.h]: s/conts/const/

File: Blocks/SelfModifyingConfig/config.h [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    // This could be atomic rename, but meh. -- D.K.
+    const std::string new_filename = HistoricalFilename(filename_, now);
+    try {
+      current::FileSystem::WriteStringToFile(JSON(config_), new_filename.c_str());
+    } catch (const current::FileException&) {
+      CURRENT_THROW(SelfModifyingConfigWriteFileException(new_filename));
+    }
+    // Then, update the config.
+    config_ = new_config;
+    // Finally, save the curent config under the original config file name.
```
Comment (mzhurovich) [inline Blocks/SelfModifyingConfig/config.h]: s/curent/current/

**Comment by mzhurovich** [comment]
@dkorolev Unfortunately, the test doesn't work because `std::strftime` prints timezone-specific value :(

File: Blocks/SelfModifyingConfig/exceptions.h [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+#include "../port.h"
+#include "../../Bricks/exception.h"
+
+namespace current {
+
+struct SelfModifyingConfigException : Exception {
+  using Exception::Exception;
+};
+
+struct SelfModifyingConfigReadFileException : Exception {
```
Comment (mzhurovich) [inline Blocks/SelfModifyingConfig/exceptions.h]: This exception and the others below should be derived from `SelfModifyingConfigException`, I suppose.

**Pull request merged**